### ISSUE TITLE
chocolate-hexen does not run the Hexen Demo

### ISSUE THREAD

**Issue by fabiangreffrath** [chocolate-doom/chocolate-doom#504]
Hi there,

unfortunately, chocolate-hexen crashes instantly if one attempts to run the Hexen Demo. This is sad, since the demo is the only version of Hexen that is freely available to get an impression of the game (i.e. there is no Freehexen).

I'd volunteer to create a patch that makes chocolate-hexen recognize and run the Hexen Demo, but I cannot make sure that it will behave *exactly* as the HEXEN.EXE that accompanies the Hexen Demo release. Will such a patch get accepted (as a starting point, at least) or will this be against Chocolate Hexen's philosophy?

 - Fabian

**Comment by fragglet** [comment]
I think supporting the Hexen demo IWAD fits firmly within the Chocolate Doom philosophy.

A patch that gets things running on the demo IWAD - even if it's not perfect yet - should be fine. Something basic that works is better than nothing. You might want to take a look at nitr8's Wii port of Chocolate Hexen, as I think he might have done some of this work already.

**Comment by fabiangreffrath** [comment]
> A patch that gets things running on the demo IWAD - even if it's not perfect yet - should be fine. Something basic that works is better than nothing.

I have pushed the hexndemo branch with what I've got so far:
https://github.com/chocolate-doom/chocolate-doom/commits/hexndemo

> You might want to take a look at nitr8's Wii port of Chocolate Hexen, as I think he might have done some of this work already.

I will. :)

**Comment by fabiangreffrath** [comment]
@derek57 
Ronald, do you remember which further changes were required to get the Hexen 4-level demo running with your wii-hexen port? I think the MAXPLAYERS macro needs to get turned into a variable and decreased to 4, right? Anything else?

**Comment by fabiangreffrath** [comment]
I'd consider the hexndemo branch pretty complete now. I haven't checked yet if the original DOS Hexen Demo version has the pick-up messages for Disc of Repulsion and Chaos Device mixed up, but that should be pretty easy once I find out where to find one of both in the first 4 levels. ;)

**Comment by derek57** [comment]
Ouch, sorry i've overseen this conversation right here. I think i made the changes using booleans. 

There do exist several DEMO versions of HEXEN as well as inofficial BETA releases. The stock DEMO IWAD is 10644136 bytes in size.

In the Wii port code there is a boolean: "HEXEN_DEMO". Apply all of those changes as well as reducing MAXPLAYERS to 4 when loading the DEMO IWAD as it doesn't include support for 8 players in multiplayer and is missing the required status bar graphics. Changing player classes would most likely crash the game then.

You also need to decrease "MAX_PLAYER_STARTS" from 8 to 4 and disable the check for player starts 5 to 8 in P_MOBJ.C. Otherwise, returning from map 3, 4 and 5 will result in getting stuck inside the floor of the SEVEN PORTALS map.

I think that should be it for the stock DEMO. There are also more changes for running the RETAIL STORE BETA #3 IWAD, the MAC DEMO IWAD as well as the 4-LEVEL BETA.

**Comment by fabiangreffrath** [comment]
> I think i made the changes using booleans. 

We have a global "gamemode" variable that we check for "shareware".

> There do exist several DEMO versions of HEXEN as well as inofficial BETA releases. The stock DEMO IWAD is 10644136 bytes in size.

This is the one against which I tested my code changes so far.

> In the Wii port code there is a boolean: "HEXEN_DEMO". Apply all of those changes as well as reducing MAXPLAYERS to 4 when loading the DEMO IWAD as it doesn't include support for 8 players in multiplayer and is missing the required status bar graphics. Changing player classes would most likely crash the game then.

Not all of these changes seem to actually apply. For example. there is mention of a "puzzle menu" in your code for which I cannot find any reference in Chocolate Hexen. Also, you seem to have moved the cheat codes around in your code. 

> You also need to decrease "MAX_PLAYER_STARTS" from 8 to 4 and disable the check for player starts 5 to 8 in P_MOBJ.C. Otherwise, returning from map 3, 4 and 5 will result in getting stuck inside the floor of the SEVEN PORTALS map.

Hm, I fail to understand why this needs to be hard-coded. I mean, the check for player start 5 to 8 could simply fail if (gamemode == shareware) or not? Could you please tell me how to exactly reproduce the bug you describe?

> I think that should be it for the stock DEMO. There are also more changes for running the RETAIL STORE BETA 3 IWAD, the MAC DEMO IWAD as well as the 4-LEVEL BETA.

I believe we should only support the latest official version, i.e. the demo IWAD with the file size you posted above.We do the same for the other games already.

**Comment by fabiangreffrath** [comment]
> Hm, I fail to understand why this needs to be hard-coded. I mean, the check for player start 5 to 8 could simply fail if (gamemode == shareware) or not? Could you please tell me how to exactly reproduce the bug you describe?

@derek57 Ronald, are you reading this, or should I rather ask by e-mail?

**Comment by fragglet** [comment]
> I believe we should only support the latest official version, i.e. the demo IWAD with the file size you posted above.We do the same for the other games already.

We do have a minor hack in place to support playing with the commercial Mac Hexen IWAD, so perhaps there's precedent for supporting the demo Mac IWAD too. I'd agree that the betas aren't worth supporting.

The hexndemo branch looks like it's definitely ready to merge into the mainline. Great work! Only bug I found is that the last of the "info" pages linked from the menu is shown as junk. I found this in mn_menu.c which may explain it:

        if (gamemode == shareware)
        {
            InfoType = (InfoType + 1) % 5;
        }
        else
        {
            InfoType = (InfoType + 1) % 4;
        }

Looks like it's been like this since the source release - 1670801d tweaked the condition expression slightly. Maybe it's the remains of a Hexen shareware release that never got finished.

**Comment by fabiangreffrath** [comment]
> The hexndemo branch looks like it's definitely ready to merge into the mainline. Great work! Only bug I found is that the last of the "info" pages linked from the menu is shown as junk.

Oh, indeed. The black-and-white stripe at the top of the screen should be the ADVISOR lump that contains the RSAC Advisory graphics; the remainder is the INTERPIC lump. I think it should be `InfoType = (InfoType + 1) % 4;` in all cases, so we could simply comment out the first branch of the `if` statement.

I have now yet been able to reproduce the bug that @derek57 mentioned and the IWAD demos fail miserably. Not sure if we should merge nevertheless or iron out these rough edges before.

**Comment by fragglet** [comment]
I'll leave it up to you to decide, but I'm certainly happy with merging it in its current state.

**Comment by fabiangreffrath** [comment]
I have fixed playback of the IWAD demos with one of the latest commits.

In this commit, which is quite intrusive, I changed each occurrence of MAXPLAYERS, which is hard-coded to 8, to a global variable "maxplayers" which is either 8 or 4 depending on (gamemode == shareware)" or not. This seems the correct approach to me, as apparently the value was hard-coded to 4 in the Demo release and later increased to eight before the release of the full version.

i think I have also found the bug @derek57 referred to, which seems to be caused by him hard-coding the value of MAXPLAYERS back to 4 instead of handling its value depending on gamemode.

The "info" menu has been fixed by commenting out the faulty if-branch.

I don't know of any other issues, so it would really help if this code would get some wider testing in the master branch. ;)
